### PR TITLE
Reduce redundancy in sort_enforcement tests

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

re https://github.com/apache/arrow-datafusion/issues/4943

# Rationale for this change

Rationale: I am working on a bug related to sort enforcement and wanted to add a new test and the current structure was very hard to do without a lot of copy / paste.

# What changes are included in this PR?

refactor repetition in sort_enforcement.rs tests into a macro

# Are these changes tested?
Only tests

# Are there any user-facing changes?
No -- there is no intended change in behavior

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 1]
```diff
@@ -589,71 +589,84 @@ mod tests {
...
-            .to_string();
-        let expected = { vec!["SortExec: [nullable_col@0 ASC]"] };
-        let actual: Vec<&str> = formatted.trim().lines().collect();
-        let actual_len = actual.len();
-        let actual_trim_last = &actual[..actual_len - 1];
-        assert_eq!(
-            expected, actual_trim_last,
-            "\n\nexpected:\n\n{expected:#?}\nactual:\n\n{actual:#?}\n\n"
-        );
+        let source = memory_exec(&schema);
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:638]: I think the new structure is much clearer about what the tests are doing -- if you look at the whitespace blind diff https://github.com/apache/arrow-datafusion/pull/4928/files?w=1 you can see that all the plans (original and optimized) are the same

**Comment by alamb** [comment]
cc @mustafasrepo and @mingmwang  who I think have worked on this code

**Comment by mustafasrepo** [comment]
This is much cleaner and more readable. Thanks @alamb.

**Comment by mingmwang** [comment]
Nice !!

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 2]
```diff
@@ -699,214 +713,147 @@ mod tests {
...
+    }
+
+    fn sort_preserving_merge_exec(
+        sort_exprs: impl IntoIterator<Item = PhysicalSortExpr>,
+        input: Arc<dyn ExecutionPlan>,
+    ) -> Arc<dyn ExecutionPlan> {
+        let sort_exprs = sort_exprs.into_iter().collect();
+        Arc::new(SortPreservingMergeExec::new(sort_exprs, input))
+    }
+
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:852]: Maybe we can add here one more function to encapsulate window exec creation
```rust
    fn window_exec(
        input: Arc<dyn ExecutionPlan>,
        schema: SchemaRef,
        sort_exprs: &[PhysicalSortExpr],
        fn_arg_column_name: &str,
    ) -> Result<Arc<dyn ExecutionPlan>> {
        let fn_arg = col(fn_arg_column_name, &schema)?;
        Ok(Arc::new(WindowAggExec::try_new(
            vec![create_window_expr(
                &WindowFunction::AggregateFunction(AggregateFunction::Count),
                "count".to_owned(),
                &[fn_arg],
                &[],
                sort_exprs,
                Arc::new(WindowFrame::new(true)),
                schema.as_ref(),
            )?],
            input,
            schema,
            vec![],
            Some(sort_exprs.to_vec()),
        )?) as Arc<dyn ExecutionPlan>)
    }
```

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 3]
```diff
@@ -664,32 +677,33 @@ mod tests {
...
                 col("non_nullable_col", schema.as_ref()).unwrap(),
             )),
-            sort_exec,
-        )?) as Arc<dyn ExecutionPlan>;
+            sort,
+        );
+
         // let filter_exec = sort_exec;
-        let window_agg_exec = Arc::new(WindowAggExec::try_new(
+        let physical_plan = Arc::new(WindowAggExec::try_new(
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:706]: given util function below is available we can construct physical plan with
``` rust
        let physical_plan = window_exec(
            filter.clone(),
            filter.schema(),
            &sort_exprs,
            "non_nullable_col",
        )?;
```

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 4]
```diff
@@ -589,71 +589,84 @@ mod tests {
...
                 descending: true,
                 nulls_first: true,
             },
-        }];
-        let sort_exec = Arc::new(SortExec::try_new(sort_exprs.clone(), source, None)?)
-            as Arc<dyn ExecutionPlan>;
+        )];
+        let sort = sort_exec(sort_exprs.clone(), source);
+
         let window_agg_exec = Arc::new(WindowAggExec::try_new(
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:670]: given util function `window_exec` is available, we can use below snippet to create `window_agg_exec`
```rust
let window_agg_exec =
    window_exec(sort.clone(), sort.schema(), &sort_exprs, "non_nullable_col")?;
```

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs (hunk 2 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/sort_enforcement.rs:852]: By the way this is just a suggestion. I think, with or without this change PR is ready to merge

**Comment by alamb** [comment]
Update https://github.com/apache/arrow-datafusion/issues/4943 is the bug I am working on with sort enforcement

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs (hunk 4 above)
Comment (alamb) [reply datafusion/core/src/physical_optimizer/sort_enforcement.rs:670]: Thank you -- I was being lazy -- I will do so

**Comment by alamb** [comment]
Will implement suggestions in a  follow on PR

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs (hunk 4 above)
Comment (alamb) [reply datafusion/core/src/physical_optimizer/sort_enforcement.rs:670]: in https://github.com/apache/arrow-datafusion/pull/4945

**Pull request merged**